storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,328 @@
+	if w.segment == nil {
+		w.segment, err = wal.OpenReadSegment(wal.SegmentName(w.walDir, w.currentSegment))
+	}
+	if err != nil {
+		w.logger.Log("err", err)
+		return
+	}
+	defer func() {
+		w.segment.Close()
+		w.segment = nil
+	}()
+
+	for {
+		select {
+		case <-w.quit:
+			return
+		// TODO: callum, handle maintaining the WAL pointer somehow across apply configs?
+		case event, ok := <-w.watcher.Events:
+			_, fileName := filepath.Split(event.Name)
+
+			if !ok {
+				// TODO: callum, in what case do we get !ok?
+				return
+			}
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				if fileName != currentSegmentName {
+					continue
+				}
+				w.decodeSegment(w.segment)
+			}
+			if event.Op&fsnotify.Create == fsnotify.Create {
+				if ok := segmentRegex.MatchString(fileName); ok {
+					// TODO: callum, ensure the create was for the segment n+1 and not n+2 or more
+					// Read the current segment one last time in case we receive events out of order. Not sure if this is necessary.
+					return
+				}
+				if ok := checkpointRegex.MatchString(fileName); ok {
+					w.logger.Log("msg", "checkpoint created", "checkpoint", event.Name)
+					// Head was truncated and WAL segments were checkpointed, so we should trim older series
+					w.writer.TrimSeries()
+				}
+			}
+		case err, ok := <-w.watcher.Errors:
+			// TODO: callum, are these errors recoverable?
+			if !ok {
+				return
+			}
+			level.Error(w.logger).Log("err", err)
+		}
+	}
+}
+
+func (w *WALWatcher) runWatcher() {
+	// The WAL dir may not exist when Prometheus first starts up.
+	for {
+		if _, err := os.Stat(w.walDir); os.IsNotExist(err) {
+			time.Sleep(time.Second)
+		} else {
+			break
+		}
+	}
+
+	var err error
+	nw, err := wal.New(nil, nil, w.walDir)
+	if err != nil {
+		w.logger.Log("err", err)
+		return
+	}
+
+	_, n, err := nw.Segments()
+	if err != nil {
+		w.logger.Log("err", err)
+		return
+	}
+
+	if n == -1 {
+		w.logger.Log("err", "no segments found")
+		return
+	}
+
+	w.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		w.logger.Log("err", err)
+		return
+	}
+
+	defer w.watcher.Close()
+	// Read series records in the current WAL and latest checkpoint, get the segment pointer back.
+	w.segment, err = w.readToEnd(w.walDir, n)
+	if err != nil {
+		level.Error(w.logger).Log("err", err)
+		return
+	}
+
+	w.currentSegment = n
+
+	err = w.watcher.Add(w.walDir)
+	if err != nil {
+		level.Error(w.logger).Log("err", err)
+		return
+	}
+
+	for {
+		select {
+		case <-w.quit:
+			return
+		default:
+			w.logger.Log("msg", "watching segment", "segment", w.currentSegment)
+			w.watch()
+			w.currentSegment++
+		}
+	}
+}
+
+// Blocks  until the sample is sent to all remote write endpoints
+func (w *WALWatcher) forwardSamples(s []tsdb.RefSample) error {
+	for {
+		select {
+		case <-w.quit:
+			return fmt.Errorf("exit when attempting to forward samples")
+		default:
+			ok := w.writer.Append(s)
```

Comment:
Now that Append retries forever, does it make sense to return an error?  If not, this whole loop can go away.